soci/soci_index.go

```diff
@@ -94,6 +112,60 @@ type IndexDescriptorInfo struct {
 	CreatedAt time.Time
 }
 
+// DecodeIndex deserializes a JSON blob in an io.Reader
+// into a SOCI index. The blob can be either an OCI 1.1 Artifact
+// or an OCI 1.0 Manifest
+func DecodeIndex(r io.Reader, index *Index) error {
+	b, err := io.ReadAll(r)
+	if err != nil {
+		return err
+	}
+	return UnmarshalIndex(b, index)
+}
+
+// UnmarshalIndex deserializes a JSON blob in a byte array
+// into a SOCI index. The blob can be either an OCI 1.1 Artifact
+// or an OCI 1.0 Manifest
+func UnmarshalIndex(b []byte, index *Index) error {
+	err := json.Unmarshal(b, index)
+	if err == nil && index.MediaType == ocispec.MediaTypeArtifactManifest {
+		return nil
+	}
+	var manifest ocispec.Manifest
+	err = json.Unmarshal(b, &manifest)
+	if err == nil {
+		fromManifest(manifest, index)
+		return nil
+	}
+	return err
+}
+
+// fromManifest converts and OCI 1.0 Manifest to a SOCI Index
+func fromManifest(manifest ocispec.Manifest, index *Index) {
+	index.MediaType = manifest.MediaType
+	index.ArtifactType = SociIndexArtifactType
+	index.Blobs = manifest.Layers
+	index.Subject = manifest.Subject
+	index.Annotations = manifest.Annotations
+}
+
+// Marshal serializes a SOCI index into a JSON blob.
+// The JSON blob will be either an OCI 1.1 Artifact or
+// an OCI 1.0 Manifest, depending on how the index was created.
+func (i *Index) Marshal() ([]byte, error) {
+	if i.MediaType == OCIArtifactManifestMediaType {
+		return json.Marshal(i)
+	}
+	var manifest ocispec.Manifest
+	manifest.SchemaVersion = 2
+	manifest.MediaType = ocispec.MediaTypeImageManifest
+	manifest.Config = defaultConfigDescriptor
+	manifest.Layers = i.Blobs
+	manifest.Subject = i.Subject
+	manifest.Annotations = i.Annotations
+	return json.Marshal(manifest)
+}
```

Comment:
Should we change the func signature to

```golang
func MarshalIndex(i *Index) ([]byte, error)
```

to be consistent with `UnmarshalIndex`.

For `ztoc`, we also have `func Unmarshal(serializedZtoc io.Reader) (*Ztoc, error)` and `func Marshal(ztoc *Ztoc) (io.Reader, ocispec.Descriptor, error)` (i.e., not `func (ztoc *Ztoc) Marshal() (io.Reader, ocispec.Descriptor, error)`).

Reply:
I guess so. I'm usually not a huge fan of package functions that operation on objects because it encourages thinking procedurally which was one of the problems we had with the index builder until we refactored it.

That said, serialization is probably not a property of the object itself and we previously removed the concept of a ztocmarshaller because it wasn't adding any value. I think this is the right change here, but I'm not totally sold on this pattern in general.